[MUI v5] Type Error when using Icons from @mui/icons-material

### 📜 Description

When upgrading to MUI v5 we are getting a type error when using Icons from `@mui/icons-material`. We see this in the following places:

- The `icon` prop of the `SidebarItem`, `SidebarSubmenuItem`, and `MyGroupsSidebarItem` components
- The `UNSTABLE_extraContextMenuItems` prop of the `EntityLayout` component
- When passing in icons to the `icons` section of `createApp` in the `App.tsx`

The error is:

```text
Type 'OverridableComponent<SvgIconTypeMap<{}, "svg">> & { muiName: string; }' is not assignable to type 'ComponentType<{ fontSize?: "default" | "large" | "small" | undefined; }>'.
  Type 'OverridableComponent<SvgIconTypeMap<{}, "svg">> & { muiName: string; }' is not assignable to type 'FunctionComponent<{ fontSize?: "default" | "large" | "small" | undefined; }>'.
    Types of parameters 'props' and 'props' are incompatible.
      Type 'PropsWithChildren<{ fontSize?: "default" | "large" | "small" | undefined; }>' is not assignable to type '{ component: any; } & { children?: ReactNode; classes?: Partial<SvgIconClasses> | undefined; color?: OverridableStringUnion<"inherit" | "disabled" | "primary" | ... 5 more ... | "warning", SvgIconPropsColorOverrides> | undefined; ... 6 more ...; viewBox?: string | undefined; } & CommonProps & Omit<...>'.
        Property 'component' is missing in type 'PropsWithChildren<{ fontSize?: "default" | "large" | "small" | undefined; }>' but required in type '{ component: any; }'.t
```

### 👍 Expected behavior

You should be able to use icons from `@mui/icons-material` without any type errors

### 👎 Actual Behavior with Screenshots

We get this error:

```text
Type 'OverridableComponent<SvgIconTypeMap<{}, "svg">> & { muiName: string; }' is not assignable to type 'ComponentType<{ fontSize?: "default" | "large" | "small" | undefined; }>'.
  Type 'OverridableComponent<SvgIconTypeMap<{}, "svg">> & { muiName: string; }' is not assignable to type 'FunctionComponent<{ fontSize?: "default" | "large" | "small" | undefined; }>'.
    Types of parameters 'props' and 'props' are incompatible.
      Type 'PropsWithChildren<{ fontSize?: "default" | "large" | "small" | undefined; }>' is not assignable to type '{ component: any; } & { children?: ReactNode; classes?: Partial<SvgIconClasses> | undefined; color?: OverridableStringUnion<"inherit" | "disabled" | "primary" | ... 5 more ... | "warning", SvgIconPropsColorOverrides> | undefined; ... 6 more ...; viewBox?: string | undefined; } & CommonProps & Omit<...>'.
        Property 'component' is missing in type 'PropsWithChildren<{ fontSize?: "default" | "large" | "small" | undefined; }>' but required in type '{ component: any; }'.t
```

### 👟 Reproduction steps

To keep things simple I'm going to base this on the main Backstage repo but you can easily do the same with a fresh instance created with `npx @backstage/create-app@latest`

1. Add `"@mui/icons-material": "^5.11.16"` to the `package/app/package.json`
2. Run `yarn install`
3. In the `packages/app/src/components/Root/Root.tsx` change this import `import HomeIcon from '@material-ui/icons/Home';` to be like this `import HomeIcon from '@mui/icons-material/Home';`
4. Run `yarn tsc`

Notice: You'll get a type error

### 📃 Provide the context for the Bug.

While upgrading to MUI v5 we ran into a number of cases like this. Now we can work around this by adding `as IconComponent` but we feel this should probably been something that is resolved in the Backstage itself as it would make this upgrade smoother and with less manual changes

### 🖥️ Your Environment

This really isn't an environment specific issue more around the usage of `@mui/icons-material`

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

In the process of upgrading and running into the same issue as well.

```
$ tsc
packages/app/src/App.tsx:101:9 - error TS2322: Type 'OverridableComponent<SvgIconTypeMap<{}, "svg">> & { muiName: string; }' is not assignable to type 'IconComponent'.

101         code: CodeIcon,
            ~~~~

packages/app/src/App.tsx:102:9 - error TS2322: Type 'OverridableComponent<SvgIconTypeMap<{}, "svg">> & { muiName: string; }' is not assignable to type 'IconComponent'.

102         alarm: AlarmIcon,
            ~~~~~
```

--- a/.changeset/thirty-pets-fry.md
+++ b/.changeset/thirty-pets-fry.md
@@ -1,0 +1,8 @@
+---
+'@backstage/frontend-plugin-api': minor
+'@backstage/core-plugin-api': minor
+---
+
+**BREAKING PRODUCERS**: The `IconComponent` no longer accepts `fontSize="default"`. This has effectively been removed from Material-UI since its last two major versions, and has not worked properly for them in a long time.
+
+This change should not have an effect on neither users of MUI4 nor MUI5/6, since the updated interface should still let you send the respective `SvgIcon` types into interfaces where relevant (e.g. as app icons).

--- a/packages/core-plugin-api/report.api.md
+++ b/packages/core-plugin-api/report.api.md
@@ -505,14 +505,9 @@ export const googleAuthApiRef: ApiRef<
 >;
 
 // @public
-export type IconComponent = ComponentType<
-  | {
-      fontSize?: 'large' | 'small' | 'default' | 'inherit';
-    }
-  | {
-      fontSize?: 'medium' | 'large' | 'small' | 'inherit';
-    }
->;
+export type IconComponent = ComponentType<{
+  fontSize?: 'medium' | 'large' | 'small' | 'inherit';
+}>;
 
 // @public
 export type IdentityApi = {

--- a/packages/core-plugin-api/src/icons/types.test.ts
+++ b/packages/core-plugin-api/src/icons/types.test.ts
@@ -1,0 +1,73 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as React from 'react';
+import { type IconComponent } from './types';
+
+// Emulate the MUI4 icon type
+type Mui4Icon = (props: {
+  fontSize?: 'default' | 'inherit' | 'large' | 'medium' | 'small';
+  otherProp?: string;
+}) => JSX.Element;
+
+// Emulate the MUI5 icon type
+type Mui5Icon = (props: {
+  fontSize?: 'inherit' | 'large' | 'medium' | 'small';
+  otherProp?: string;
+}) => JSX.Element;
+
+// Emulate the MUI6 icon type
+interface OverridableComponent<P> {
+  <RootComponent extends React.ElementType>(
+    props: {
+      /**
+       * The component used for the root node.
+       * Either a string to use a HTML element or a component.
+       */
+      component: RootComponent;
+    } & { stuff: number },
+  ): React.JSX.Element | null;
+  (props: P): React.JSX.Element | null;
+}
+type Mui6Icon = OverridableComponent<{
+  fontSize?: 'inherit' | 'large' | 'medium' | 'small';
+  otherProp?: string;
+}>;
+
+type NotAnIcon1 = (props: {
+  fontSize?: 'foo';
+  otherProp?: string;
+}) => JSX.Element;
+
+type NotAnIcon2 = (props: {
+  fontSize?: number;
+  otherProp?: string;
+}) => JSX.Element;
+
+describe('IconComponent', () => {
+  // eslint-disable-next-line jest/expect-expect
+  it('should be a component type', () => {
+    // @ts-ignore
+    let icon: IconComponent;
+    icon = {} as Mui4Icon;
+    icon = {} as Mui5Icon;
+    icon = {} as Mui6Icon;
+    // @ts-expect-error
+    icon = {} as NotAnIcon1;
+    // @ts-expect-error
+    icon = {} as NotAnIcon2;
+  });
+});

--- a/packages/core-plugin-api/src/icons/types.ts
+++ b/packages/core-plugin-api/src/icons/types.ts
@@ -22,7 +22,7 @@ import { ComponentType } from 'react';
  *
  * @remarks
  *
- * The type is based on SvgIcon from Material UI, but both do not what the plugin-api
+ * The type is based on SvgIcon from Material UI, but we do not want the plugin-api
  * package to have a dependency on Material UI, nor do we want the props to be as broad
  * as the SvgIconProps interface.
  *
@@ -32,14 +32,6 @@ import { ComponentType } from 'react';
  *
  * @public
  */
-
-export type IconComponent = ComponentType<
-  /* Material UI v4 */
-  | {
-      fontSize?: 'large' | 'small' | 'default' | 'inherit';
-    }
-  /* Material UI v5: https://mui.com/material-ui/migration/v5-component-changes/#icon */
-  | {
-      fontSize?: 'medium' | 'large' | 'small' | 'inherit';
-    }
->;
+export type IconComponent = ComponentType<{
+  fontSize?: 'medium' | 'large' | 'small' | 'inherit';
+}>;

--- a/packages/frontend-plugin-api/report.api.md
+++ b/packages/frontend-plugin-api/report.api.md
@@ -1285,14 +1285,9 @@ export const IconBundleBlueprint: ExtensionBlueprint<{
 }>;
 
 // @public
-export type IconComponent = ComponentType<
-  | {
-      fontSize?: 'large' | 'small' | 'default' | 'inherit';
-    }
-  | {
-      fontSize?: 'medium' | 'large' | 'small' | 'inherit';
-    }
->;
+export type IconComponent = ComponentType<{
+  fontSize?: 'medium' | 'large' | 'small' | 'inherit';
+}>;
 
 // @public
 export interface IconsApi {

--- a/packages/frontend-plugin-api/src/icons/types.test.ts
+++ b/packages/frontend-plugin-api/src/icons/types.test.ts
@@ -1,0 +1,73 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as React from 'react';
+import { type IconComponent } from './types';
+
+// Emulate the MUI4 icon type
+type Mui4Icon = (props: {
+  fontSize?: 'default' | 'inherit' | 'large' | 'medium' | 'small';
+  otherProp?: string;
+}) => JSX.Element;
+
+// Emulate the MUI5 icon type
+type Mui5Icon = (props: {
+  fontSize?: 'inherit' | 'large' | 'medium' | 'small';
+  otherProp?: string;
+}) => JSX.Element;
+
+// Emulate the MUI6 icon type
+interface OverridableComponent<P> {
+  <RootComponent extends React.ElementType>(
+    props: {
+      /**
+       * The component used for the root node.
+       * Either a string to use a HTML element or a component.
+       */
+      component: RootComponent;
+    } & { stuff: number },
+  ): React.JSX.Element | null;
+  (props: P): React.JSX.Element | null;
+}
+type Mui6Icon = OverridableComponent<{
+  fontSize?: 'inherit' | 'large' | 'medium' | 'small';
+  otherProp?: string;
+}>;
+
+type NotAnIcon1 = (props: {
+  fontSize?: 'foo';
+  otherProp?: string;
+}) => JSX.Element;
+
+type NotAnIcon2 = (props: {
+  fontSize?: number;
+  otherProp?: string;
+}) => JSX.Element;
+
+describe('IconComponent', () => {
+  // eslint-disable-next-line jest/expect-expect
+  it('should be a component type', () => {
+    // @ts-ignore
+    let icon: IconComponent;
+    icon = {} as Mui4Icon;
+    icon = {} as Mui5Icon;
+    icon = {} as Mui6Icon;
+    // @ts-expect-error
+    icon = {} as NotAnIcon1;
+    // @ts-expect-error
+    icon = {} as NotAnIcon2;
+  });
+});

--- a/packages/frontend-plugin-api/src/icons/types.ts
+++ b/packages/frontend-plugin-api/src/icons/types.ts
@@ -22,7 +22,7 @@ import { ComponentType } from 'react';
  *
  * @remarks
  *
- * The type is based on SvgIcon from Material UI, but both do not what the plugin-api
+ * The type is based on SvgIcon from Material UI, but we do not want the plugin-api
  * package to have a dependency on Material UI, nor do we want the props to be as broad
  * as the SvgIconProps interface.
  *
@@ -32,14 +32,6 @@ import { ComponentType } from 'react';
  *
  * @public
  */
-
-export type IconComponent = ComponentType<
-  /* Material UI v4 */
-  | {
-      fontSize?: 'large' | 'small' | 'default' | 'inherit';
-    }
-  /* Material UI v5: https://mui.com/material-ui/migration/v5-component-changes/#icon */
-  | {
-      fontSize?: 'medium' | 'large' | 'small' | 'inherit';
-    }
->;
+export type IconComponent = ComponentType<{
+  fontSize?: 'medium' | 'large' | 'small' | 'inherit';
+}>;

